File: AwsCryptographicMaterialProviders/runtimes/java/src/test/LocalCMCTests.java
import Random_Compile.ExternRandom;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicIntegerArray;
import org.testng.Assert;
import org.testng.annotations.Test;
import software.amazon.cryptography.keystore.model.BeaconKeyMaterials;
import software.amazon.cryptography.materialproviders.ICryptographicMaterialsCache;
import software.amazon.cryptography.materialproviders.MaterialProviders;
import software.amazon.cryptography.materialproviders.model.CacheType;
import software.amazon.cryptography.materialproviders.model.CreateCryptographicMaterialsCacheInput;
import software.amazon.cryptography.materialproviders.model.DefaultCache;
import software.amazon.cryptography.materialproviders.model.EntryDoesNotExist;
import software.amazon.cryptography.materialproviders.model.GetCacheEntryInput;
import software.amazon.cryptography.materialproviders.model.GetCacheEntryOutput;
import software.amazon.cryptography.materialproviders.model.MaterialProvidersConfig;
import software.amazon.cryptography.materialproviders.model.Materials;
import software.amazon.cryptography.materialproviders.model.MultiThreadedCache;
import software.amazon.cryptography.materialproviders.model.PutCacheEntryInput;

public class LocalCMCTests {

  private static final ICryptographicMaterialsCache test = MaterialProviders
    .builder()
    .MaterialProvidersConfig(MaterialProvidersConfig.builder().build())
    .build()
    .CreateCryptographicMaterialsCache(
      CreateCryptographicMaterialsCacheInput
        .builder()
        .cache(
          CacheType
            .builder()
            .MultiThreaded(
              MultiThreadedCache.builder().entryCapacity(10).build()
            )
            .build()
        )
        .build()
    );
  private static final List<String> identifies = Collections.unmodifiableList(
    Arrays.asList(
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen",
      "twenty",
      "twenty one"
    )
  );
  private static final int IDS_SIZE = identifies.size();
  private static final AtomicIntegerArray CacheHitCounter =
    new AtomicIntegerArray(IDS_SIZE);
  private static final AtomicIntegerArray CacheMissCounter =
    new AtomicIntegerArray(IDS_SIZE);
  private static final AtomicIntegerArray IdCounter = new AtomicIntegerArray(
    IDS_SIZE
  );

  @Test(threadPoolSize = 10, invocationCount = 300000, timeOut = 10000)
  public void TestALotOfAdding() {
    Random rand = ExternRandom.getSecureRandom();
    int index = rand.nextInt(IDS_SIZE);
    IdCounter.getAndAdd(index, 1);
    String beaconKeyIdentifier = identifies.get(index);

    ByteBuffer cacheIdentifier = ByteBuffer.wrap(
      beaconKeyIdentifier.getBytes(StandardCharsets.UTF_8)
    );

    GetCacheEntryInput getCacheEntryInput = GetCacheEntryInput
      .builder()
      .identifier(cacheIdentifier)
      .build();

    try {
      GetCacheEntryOutput getCacheEntryOutput = test.GetCacheEntry(
        getCacheEntryInput
      );
      Assert.assertNotNull(
        getCacheEntryOutput.materials().BeaconKey(),
        "Beacon Materials are null!"
      );
      Assert.assertEquals(
        getCacheEntryOutput.materials().BeaconKey().beaconKey(),
        cacheIdentifier
      );
      CacheHitCounter.getAndAdd(index, 1);
    } catch (EntryDoesNotExist ex) {
      CacheMissCounter.getAndAdd(index, 1);
      Materials materials = Materials
        .builder()
        .BeaconKey(
          BeaconKeyMaterials
            .builder()
            .beaconKeyIdentifier(beaconKeyIdentifier)
            // The cacheIdentifier is used as the material
            // because we are not testing the cryptography here.
            .beaconKey(cacheIdentifier)
            .encryptionContext(Collections.emptyMap())
            .build()
        )
        .build();

      PutCacheEntryInput putCacheEntryInput = PutCacheEntryInput
        .builder()
        .identifier(cacheIdentifier)
        .creationTime(Instant.now().getEpochSecond())
        .expiryTime(Instant.now().getEpochSecond() + 1)
        .materials(materials)
        .build();
      test.PutCacheEntry(putCacheEntryInput);
    }
  }

  // This test asserts that there were cache misses and cache hits.
  @Test(dependsOnMethods = { "TestALotOfAdding" })
  public void validateTestALotOfAdding() {
    int sumOfHits = 0;
    int sumOfMisses = 0;
    int idSelectedCount = -1;
    for (int i = 0; i < IDS_SIZE; i++) {
      // Was the ID ever randomly selected?
      idSelectedCount = IdCounter.get(i);
      if (idSelectedCount > 1) {
        // if the ID was hit, it MUST have been a HIT or MISS
        Assert.assertEquals(
          idSelectedCount,
          CacheHitCounter.get(i) + CacheMissCounter.get(i),
          "Cache hits for ID + Cache misses for ID != ID selection"
        );
        System.out.printf(
          "LocalCMC Test: ID: %s hit %s miss %s%n",
          i,
          CacheHitCounter.get(i),
          CacheMissCounter.get(i)
        );
      }
      sumOfHits += CacheHitCounter.get(i);
      sumOfMisses += CacheMissCounter.get(i);
      idSelectedCount = -1;
    }
    // Assert there were misses, as the cache started empty
    Assert.assertTrue(sumOfMisses > 0, "How were there no misses!");
    // Assert there was at least one hit
    Assert.assertTrue(sumOfHits > 0, "How were there no hits!");
    // Without better ID selection control,
    // along with finer timing control,
    // we cannot assert expected hits or misses
  }
}

File: AwsCryptographicMaterialProviders/runtimes/rust/tests/local_cmc_tests.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

#![deny(warnings, unconditional_panic)]
#![deny(nonstandard_style)]
#![deny(clippy::all)]

pub struct BoxError(String);
pub struct BoxError2(String);

impl std::fmt::Debug for BoxError2 {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<BoxError> for BoxError2 {
    fn from(error: BoxError) -> Self {
        BoxError2(error.0)
    }
}

impl<T: std::fmt::Debug> From<T> for BoxError {
    fn from(error: T) -> Self {
        let my_str = format!("{error:?}");
        BoxError(my_str)
    }
}

pub mod local_cmc_tests {
    use aws_mpl_legacy::client as mpl_client;
    use aws_mpl_legacy::deps::aws_cryptography_keyStore::types::BeaconKeyMaterials;
    use aws_mpl_legacy::operation::get_cache_entry::GetCacheEntryOutput;
    use aws_mpl_legacy::types::cryptographic_materials_cache::CryptographicMaterialsCacheRef;
    use aws_mpl_legacy::types::error::Error;
    use aws_mpl_legacy::types::material_providers_config::MaterialProvidersConfig;
    use aws_mpl_legacy::types::CacheType;
    use aws_mpl_legacy::types::DefaultCache;
    use aws_mpl_legacy::types::Materials;
    use chrono::Utc;
    use futures::stream::StreamExt;
    use rand::{rng, Rng};
    use std::collections::HashMap;
    // Currently this is commented out to keep it consistent with other
    // runtimes but every language should eventually have a network
    // call delay while testing adding to the StormTrackingCache.
    // use std::time::Duration;

    async fn work_for_thread() -> Result<(), crate::BoxError> {
        let mpl_config: MaterialProvidersConfig = MaterialProvidersConfig::builder().build()?;
        let mpl: mpl_client::Client = mpl_client::Client::from_conf(mpl_config)?;

        let test: CryptographicMaterialsCacheRef = mpl
            .create_cryptographic_materials_cache()
            .cache(CacheType::Default(
                DefaultCache::builder().entry_capacity(10).build()?,
            ))
            .send()
            .await?;

        let identifiers: Vec<&str> = vec![
            "one",
            "two",
            "three",
            "four",
            "five",
            "six",
            "seven",
            "eight",
            "nine",
            "ten",
            "eleven",
            "twelve",
            "thirteen",
            "fourteen",
            "fifteen",
            "sixteen",
            "seventeen",
            "eighteen",
            "nineteen",
            "twenty",
            "twenty one",
        ];

        let id_size: usize = identifiers.len();

        let invocations = 300000;
        let concurrent_tasks = 10;

        futures::stream::iter(0..invocations)
            .map(|_i| {
                let identifiers = identifiers.clone();
                let test = test.clone();

                async move {
                    let random = rng().random_range(0..id_size);
                    let beacon_key_identifier: &str = identifiers[random];

                    let cache_identifier: aws_smithy_types::Blob =
                        aws_smithy_types::Blob::new(beacon_key_identifier.as_bytes());

                    let cache_entry_output: Result<GetCacheEntryOutput, Error> = test
                        .get_cache_entry()
                        .identifier(cache_identifier.clone())
                        .send()
                        .await;

                    match cache_entry_output {
                        Ok(_) => {
                            // Cache hit
                        }
                        Err(Error::EntryDoesNotExist { message: _m }) => {
                            let materials = Materials::BeaconKey(
                                BeaconKeyMaterials::builder()
                                    .beacon_key_identifier(beacon_key_identifier.to_string())
                                    // The cacheIdentifier is used as the material
                                    // because we are not testing the cryptography here.
                                    .beacon_key(cache_identifier.clone())
                                    .encryption_context(HashMap::new())
                                    .build()?,
                            );
                            // This sleep time is to mimic a KMS or DDB call
                            // Currently this is commented out to keep it consistent with other
                            // runtimes but every language should eventually have a network
                            // call delay while testing adding to the StormTrackingCache.
                            // The reason why this is commented out and not just set to 0 is because
                            // setting it to 0 adds a significant overhead of task switching, and
                            // with that, testing for 300_000 instances will not be possible.
                            // let network_call_delay: u64 = 50;
                            // tokio::time::sleep(Duration::from_millis(network_call_delay)).await;

                            test.put_cache_entry()
                                .identifier(cache_identifier)
                                .creation_time(Utc::now().timestamp())
                                .expiry_time(Utc::now().timestamp() + 1)
                                .materials(materials)
                                .send()
                                .await?;

                        }
                        Err(e) => {
                            panic!("Unexpected error while accessing cache: {e}");
                        }
                    }

                    Ok::<(), crate::BoxError>(())
                }
            })
            .buffer_unordered(concurrent_tasks)
            .collect::<Vec<_>>()
            .await;

        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_a_lot_of_adding() -> Result<(), crate::BoxError2> {
        work_for_thread().await?;
        Ok(())
    }
}
